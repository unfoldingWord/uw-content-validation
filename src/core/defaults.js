export const DEFAULT_EXCERPT_LENGTH = 20;

export const REPO_CODES_LIST = [
  'UHB', 'UGNT',
  'LT', 'ST',
  'TA', 'TW', 'TWL',
  // The ones with 2 suffix are the 2021 new TSV format repos
  'TN', 'TN2', 'TQ', 'TQ2',
  'SN', 'SQ',
  'UGL', 'UHAL',
  'OBS', 'OBS-TWL',
  'OBS-TN', 'OBS-TN2', 'OBS-TQ', 'OBS-TQ2',
  'OBS-SN', 'OBS-SQ',
];
