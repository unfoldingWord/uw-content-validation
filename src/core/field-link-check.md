## Basic Link Field Check Sandbox

This function is for checking text fields that are links, or that contain links.

```js
// The control code in this box is editable for changing settings—
//    simply click inside here and add, change, or delete text as required.
// Note that (gray) lines starting with // are "comments", i.e., they are ignored by the software
//    so if you want to enable those lines, you must remove the // from the beginning of the line.

import { checkFieldLinks } from './field-link-check';
import { processNoticesToErrorsWarnings } from '../demos/notice-processing-functions';
import { RenderRawResults, RenderSuccessesErrorsWarnings } from '../demos/RenderProcessedResults';
import { userLog } from './utilities';

// Empty, space, link, RC, good, and bad text samples
const textE = "";
const textS = " ";
const textL = "https://unfoldingWord.org";
const textRC1 = "rc://en/obs/book/obs/02/03";
const textRC2 = "rc://en/tw/dict/bible/other/canaan";
const textMD1 = "Have a look [here](https://unfoldingWord.org/)!";
const textMD2 = " Look at [[https://unfoldingWord.org]] ";
const textMD3 = "Now  look at [[http://door43.org]] ";

// Easy for user to change text used in next line (to demonstrate differing results)
const username = 'unfoldingWord';
const chosenTextName = "textMD3";
const chosenText = textMD3;

let rawResults;

// Define our callback function
function acceptUpdatedResult(newLinkResult){
    // Update the results with later results
    userLog(`acceptUpdatedResult callback function is updating result now with ${newLinkResult.noticeList.length} notices.`);
    rawResults.noticeList = newLinkResult.noticeList;
    // Now how can we tell Styleguidist to refresh???
}

const linkOptions = {
    expectedCount: 1, // 0 = all links optional (no minimal number expected)
    allowedCount: 0, // 0 = no maximum limit
    otherTextAllowed: true, // if false, field is just the link only
    linkTypesAllowed: ['RC', 'md', 'naked',],
        // 'RC' (ResourceContainer) allows 'rc://...` links
        // 'md' allows '[this](http://here.org)' and [[http://here.org]]
        // 'naked' allows 'http:://example.org' and even 'example.org'
    checkTargets: true, // Attempt to retrieve the link(s) -- slows it down
    callbackFunction: acceptUpdatedResult, // Required to accept updated results if checking targets
};

// This function returns the results of the fast checks
//  and if specified in linkOptions, the callback will update result later with results of slower checks
rawResults = checkFieldLinks(username, 'en', '', chosenTextName, chosenText, linkOptions, 'that was supplied');
if (!rawResults.successList || !rawResults.successList.length)
    rawResults.successList = ["Done basic link checks"];

<>
<b>Check</b> "{chosenText}"<br/><br/>
<RenderRawResults results={rawResults} />
</>
```
