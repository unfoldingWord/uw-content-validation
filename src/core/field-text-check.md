## Basic Text Field Check Sandbox

This `checkTextField()` function can be passed a text (incl. markdown) field and checks for basic errors like leading/trailing spaces, bad punctuation, etc.

This generic function returns a list/array of notices, that can then be post-processed to eliminate any warning types that don’t apply to this particular type of field so we don’t flood the user with a lot of false positives.

This demonstration doesn’t display the raw notices, but rather displays the processed and formatted lists of errors and warnings.

```js
// The control code in this box is editable for changing settings—
//    simply click inside here and add, change, or delete text as required.
// Note that (gray) lines starting with // are "comments", i.e., they are ignored by the software
//    so if you want to enable those lines, you must remove the // from the beginning of the line.

import { checkTextField } from './field-text-check';
import { RenderRawResults } from '../demos/RenderProcessedResults';

// Empty, space, good, and bad, link, and RC text samples
const textE = "";
const textS = " ";
const textG = "Peace on Earth, good will to all men/people!";
const textB = " Peace  on Earth,, good will to all) men! ";
const textL = "https://unfoldingWord.org";
const textRC1 = "rc://en/obs/book/obs/02/03";
const textRC2 = "Compare with rc://en/tw/dict/bible/other/canaan";

// Just change these next two lines to change the text being used (to demonstrate differing results)
const username = 'unfoldingWord';
const chosenTextName = "textB";
const chosenText = textB;

// The third parameter is "linksAllowed"
const rawResults = checkTextField(username, 'en', 'TN', 'raw', 'Sample', chosenText, false, 'in '+chosenTextName+' that was supplied');
if (!rawResults.successList || !rawResults.successList.length)
    rawResults.successList = ["Done basic text checks"];

<>
<b>Check</b> "{chosenText}"<br/><br/>
<RenderRawResults results={rawResults} />
</>
```
