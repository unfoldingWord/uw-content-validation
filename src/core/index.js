export * from './books/';
export * from './defaults';
export * from './utilities';
export * from './getApi';
export * from './text-handling-functions';

export * from './field-text-check';
export * from './file-text-check';
export * from './lexicon-file-contents-check';
export * from './markdown-text-check';
export * from './markdown-file-contents-check';
export * from './manifest-text-check';
export * from './notes-tsv7-row-check';
export * from './notes-tsv7-table-check';
export * from './plain-text-check';
export * from './questions-tsv7-row-check';
export * from './questions-tsv7-table-check';
export * from './notes-links-check';
export * from './tn-tsv9-row-check';
export * from './tn-tsv9-table-check';
export * from './twl-tsv6-row-check';
export * from './twl-tsv6-table-check';
export * from './usfm-text-check';
export * from './yaml-text-check';
