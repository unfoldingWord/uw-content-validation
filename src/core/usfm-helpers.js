// eslint-disable-next-line no-unused-vars
import { CLOSING_PUNCTUATION_CHARACTERS, HEBREW_CANTILLATION_REGEX, PAIRED_PUNCTUATION_OPENERS, PAIRED_PUNCTUATION_CLOSERS } from './text-handling-functions';
// eslint-disable-next-line no-unused-vars
import { functionLog, debugLog, parameterAssert, logicAssert, dataAssert, ourParseInt, aboutToOverwrite } from './utilities';

export function extractTextFromComplexUSFM(usfmText) {
  // functionLog(`extractTextFromComplexUSFM(${usfmText.length} chars)…`)
  // Do initial global replaces
  usfmText = usfmText.replace(/\\zaln-e\\\*/g, ''); // Remove \zaln-e self-closed milestones
  usfmText = usfmText.replace(/\\k-e\\\*/g, ''); // Remove \k-e self-closed milestones
  // debugLog(`  extractTextFromComplexUSFM: got verse text1: (${usfmText.length}) '${usfmText}'`);
  usfmText = usfmText.replace(/\\zaln-s.+?\\\*/g, ''); // Remove \zaln-s self-closed milestones
  usfmText = usfmText.replace(/\\k-s.+?\\\*/g, ''); // Remove \k-s self-closed milestones
  // debugLog(`  extractTextFromComplexUSFM: got verse text2: (${usfmText.length}) '${usfmText}'`);
  usfmText = usfmText.replace(/\\w (.+?)\|.+?\\w\*/g, '$1'); // Replace word fields with just the word
  // debugLog(`  extractTextFromComplexUSFM: got verse text3a: (${usfmText.length}) '${usfmText}'`);
  usfmText = usfmText.replace(/{|}/g, ''); // Replace (ST) "added" markers
  // debugLog(`  extractTextFromComplexUSFM: got verse text3b: (${usfmText.length}) '${usfmText}'`);

  // Remove lines we don't want
  const hadTrailingNewline = usfmText.slice(-1) === '\n';
  const newLines = [];
  for (let line of usfmText.split('\n')) {
    if (!line.length) continue;
    if (line.startsWith('\\id ')) continue;
    if (line.startsWith('\\usfm ')) continue;
    if (line.startsWith('\\ide ')) continue;
    if (line.startsWith('\\h ')) continue;
    if (line.startsWith('\\toc')) continue;
    if (line === '\\ts\\*' || line === '\\ts\\* ') continue; // Handle known trailing spaces
    if (line.startsWith('\\rem ')) continue;
    // debugLog(`  extractTextFromComplexUSFM got line text4a: (${line.length}) '${line}'`);
    if (line.startsWith('\\c ')) line = `[C${line.slice(3)}]`;
    if (line.startsWith('\\v ')) {
      const rest = line.slice(3);
      const restBits = rest.split(' ');
      const vNum = restBits[0];
      let more = rest.slice(vNum.length + 1);
      if (more === undefined) more = '';
      if (more.length && more[0] !== ' ') more = ` ${more}`;
      line = `‹v${vNum}›${more}`;
    }
    // debugLog(`  extractTextFromComplexUSFM got line text4z: (${line.length}) '${line}'`);
    if (line.startsWith('\\p ') && line.trim() === '\\p') line = '\\p'; // Handle known trailing spaces
    newLines.push(line);
  }
  usfmText = newLines.join(' ').replace(/\\p /g, '\n')
  // debugLog(`  extractTextFromComplexUSFM got verse text5: (${usfmText.length}) '${usfmText}'`);
  if (hadTrailingNewline) usfmText += '\n';
  // debugLog(`  extractTextFromComplexUSFM got verse text5b: (${usfmText.length}) '${usfmText}'`);
  logicAssert(hadTrailingNewline === (usfmText.slice(-1) === '\n'), `extractTextFromComplexUSFM hadTrailingNewline was ${hadTrailingNewline} now different!`);

  usfmText = usfmText.replace(/ ‹v\d{1,3}›/g, ''); // Remove (now-adjusted) verse numbers inside the text so we get the flowing text

  usfmText = usfmText.replace(/ {2,}/g, ' ') // Eliminate multiple spaces
  // debugLog(`  extractTextFromComplexUSFM got verse text6: (${usfmText.length}) '${usfmText}'`);


  // Remove footnotes
  // NOTE: If there's two footnotes and no closer on the first one, this replace will swallow intervening text
  //  but this isn't the place to worry about bad USFM in the original languages
  //  so the quote match will possibly fail as a consequential error
  usfmText = usfmText.replace(/\\f (.+?)\\f\*/g, '');
  // Remove alternative versifications
  usfmText = usfmText.replace(/\\va (.+?)\\va\*/g, '');
  // debugLog(`Got verse text3: '${usfmText}'`);

  // Final clean-up (shouldn’t be necessary, but just in case)
  usfmText = usfmText.replace(/ {2,}/g, ' '); // Eliminate multiple spaces
  logicAssert(usfmText.indexOf('  ') === -1, `getOriginalVerse: Should be no double spaces left in '${usfmText}'`);
  //logicAssert(usfmText.indexOf('\\w') === -1, `getOriginalVerse: Should be no \\w in ${bookID} ${C}:${V} '${usfmText}'`);
  //logicAssert(usfmText.indexOf('\\k') === -1, `getOriginalVerse: Should be no \\k in ${bookID} ${C}:${V} '${usfmText}'`);
  //logicAssert(usfmText.indexOf('x-') === -1, `getOriginalVerse: Should be no x- in ${bookID} ${C}:${V} '${usfmText}'`);
  //logicAssert(usfmText.indexOf('\\f') === -1, `getOriginalVerse: Should be no \\f in ${bookID} ${C}:${V} '${usfmText}'`);
  //logicAssert(usfmText.indexOf('\\x') === -1, `getOriginalVerse: Should be no \\x in ${bookID} ${C}:${V} '${usfmText}'`);

  // debugLog(`  extractTextFromComplexUSFM is returning (${usfmText.length}) '${usfmText}'`);
  // debugLog(`  extractTextFromComplexUSFM is returning (${usfmText.length}) characters`);
  return usfmText;
}
