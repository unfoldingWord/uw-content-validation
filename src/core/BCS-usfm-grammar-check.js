import grammar from 'usfm-grammar';
import * as books from './books/books';
import { DEFAULT_EXCERPT_LENGTH } from './defaults'
// eslint-disable-next-line no-unused-vars
import { userLog, debugLog, functionLog, parameterAssert, ourParseInt } from './utilities';


// const USFM_GRAMMAR_VALIDATOR_VERSION_STRING = '0.4.5';

const LINE_COLUMN_NUMBERS_REGEX = new RegExp('Line (\\d{1,6}), col (\\d{1,4}):'); // e.g., "Line 1538, col 4: 1537 ..."


/**
 *
 * @param {string} strictnessString -- 'strict' or 'relaxed'
 * @param {string} bookID -- 3-character book ID
 * @param {string} fileText -- the actual USFM text
 * @param {string} filename -- for error messages
 * @param {string} givenLocation -- for error messages
 * @param {Object} checkingOptions -- optional options
 * @returns {Object} including isValidUSFM flag
 */
export function runBCSGrammarCheck(strictnessString, bookID, fileText, filename, givenLocation, checkingOptions) {
    // Runs the BCS USFM Grammar checker
    //  which can be quite time-consuming on large, complex USFM files
    // userLog(`Running ${strictnessString} BCS USFM grammar check${givenLocation} (can take quite a while for a large book)…`);
    //parameterAssert(strictnessString === 'strict' || strictnessString === 'relaxed', `Unexpected strictnessString='${strictnessString}'`);

    let excerptLength;
    try {
        excerptLength = checkingOptions?.excerptLength;
    } catch (usfmELerror) { }
    if (typeof excerptLength !== 'number' || isNaN(excerptLength)) {
        excerptLength = DEFAULT_EXCERPT_LENGTH;
        // debugLog(`Using default excerptLength=${excerptLength}`);
    }
    // else
    // debugLog(`Using supplied excerptLength=${excerptLength} cf. default=${DEFAULT_EXCERPT_LENGTH}`);
    const excerptHalfLength = Math.floor(excerptLength / 2); // rounded down
    const excerptHalfLengthPlus = Math.floor((excerptLength + 1) / 2); // rounded up
    // debugLog(`Using excerptHalfLength=${excerptHalfLength}`, `excerptHalfLengthPlus=${excerptHalfLengthPlus}`);

    // Now create the parser and run the check
    let parserToJSONResultObject, parseWarnings;
    try {
        // debugLog(`${new Date().getTime() / 1000} Setting-up the USFMGrammar checker…`);
        const ourUsfmParser = new grammar.USFMParser(fileText, strictnessString === 'relaxed' ? grammar.LEVEL.RELAXED : null);
        // Returns a Boolean indicating whether the input USFM text satisfies the grammar or not.
        // This method is available in both default and relaxed modes.
        // const parserResult = ourUsfmParser.validate();
        // debugLog(`${new Date().getTime() / 1000} Running the BCS USFMGrammar checker (may take several seconds)…`);
        // userLog(`Running the BCS USFMGrammar checker for ${bookID} (may take several seconds)…`);
        parserToJSONResultObject = ourUsfmParser.toJSON();
        // debugLog(`${new Date().getTime() / 1000} Got the BCS USFMGrammar checker toJSON result: ${Object.keys(parserToJSONResultObject)}`);
        // debugLog(`${new Date().getTime() / 1000} Got the BCS USFMGrammar checker toJSON _messages: ${Object.keys(parserToJSONResultObject._messages)}`);
        // debugLog(`${new Date().getTime() / 1000} Got the BCS USFMGrammar checker: ${Object.keys(ourUsfmParser)}`);
        parseWarnings = parserToJSONResultObject._warnings ? parserToJSONResultObject._warnings : ourUsfmParser.warnings;
        // debugLog(`${new Date().getTime() / 1000} Got warnings from the BCS USFMGrammar checker: (${parseWarnings.length}) ${parseWarnings}`);
    } catch (parserError) { // This is how the Parser returns USFM errors, i.e., it stops after the first error
        // debugLog(`${new Date().getTime() / 1000} Got an exception when using the BCS USFMGrammar checker: ${parserError}`);
        const ourErrorObject = {
            priority: 840, message: "USFMGrammar check failed",
            details: parserError,
            filename,
            location: givenLocation
        };
        try { // See if we can improve the result with line and column numbers
            // NOTE: The following code is quite fragile
            //  as it depends on the precise format of the error message returned from USFMParser
            const regexMatchObject = LINE_COLUMN_NUMBERS_REGEX.exec(parserError);
            const [totalLink, lineNumberString, columnNumberString] = regexMatchObject;
            ourErrorObject.lineNumber = ourParseInt(lineNumberString);
            ourErrorObject.characterIndex = ourParseInt(columnNumberString) - 1;
            const errorLineText = fileText.split('\n')[ourErrorObject.lineNumber - 1];
            ourErrorObject.excerpt = (ourErrorObject.characterIndex > excerptHalfLength ? '…' : '') + errorLineText.substring(ourErrorObject.characterIndex - excerptHalfLength, ourErrorObject.characterIndex + excerptHalfLengthPlus) + (ourErrorObject.characterIndex + excerptHalfLengthPlus < errorLineText.length ? '…' : '');
            // NOTE: Not 100% sure that it’s more helpful to the user if we do this next line ???
            ourErrorObject.details = ourErrorObject.details.slice(totalLink.length); // Delete the line and column numbers that we found
        } catch (secondError) {
            debugLog(`USFMGrammar second error: ${secondError}`);
        }
        // if (ourErrorObject.excerpt.startsWith('\\va ') || ourErrorObject.excerpt.startsWith('\\ca ')) // lower the priority
        if (parserError.indexOf('\\va ') >= 0 || parserError.indexOf('\\ca ') >= 0)
            ourErrorObject.priority = 140; // from 840
        return { isValidUSFM: false, error: ourErrorObject, warnings: [] };
    }
    let parserMessages;
    parserMessages = parserToJSONResultObject._messages; // Throw away the JSON (if any)
    // debugLog(`  Finished BCS USFM grammar check with messages: ${JSON.stringify(parserResult)}\n and warnings: ${JSON.stringify(ourUsfmParser.warnings)}.`);

    // TODO: I think most of the following code is now obsolete and can be deleted
    let parseError;
    parseError = parserMessages._error;
    let ourErrorMessage, lineNumberString, characterIndex, excerpt;
    // NOTE: The following code is quite fragile
    //  as it depends on the precise format of the error message returned from USFMParser
    let ourErrorObject = {};
    if (parseError) {
        debugLog("Oh! This USFMGrammar check code IS still needed!!!");
        const contextRE = /(\d+?)\s\|\s(.+)/g;
        for (const errorLine of parseError.split('\n')) {
            // debugLog(`BCS errorLine=${errorLine}`);
            if (errorLine.startsWith('>')) {
                const regexResult = contextRE.exec(errorLine.slice(1).trim());
                // debugLog(`  regexResult: ${JSON.stringify(regexResult)}`);
                if (regexResult) {
                    lineNumberString = regexResult[1];
                    excerpt = regexResult[2];
                }
            }
            else if (errorLine.endsWith('^')) {
                characterIndex = errorLine.indexOf('^') - 8;
                if (characterIndex < 0) characterIndex = 0; // Just in case
                if (excerpt.length)
                    excerpt = (characterIndex > excerptHalfLength ? '…' : '') + excerpt.substring(characterIndex - excerptHalfLength, characterIndex + excerptHalfLengthPlus) + (characterIndex + excerptHalfLengthPlus < excerpt.length ? '…' : '')
            }
            else ourErrorMessage = errorLine; // We only want the last one
        }
        // debugLog(`  ourErrorMessage: '${ourErrorMessage}' lineNumberString=${lineNumberString} characterIndex=${characterIndex} excerpt='${excerpt}'`);

        // Some of these "errors" need to be degraded in priority

        let adjustedPriority = 594; // We don’t make these extra high coz the messages are hard for users to interpret
        if (excerpt === '\\s5' // Temporarily, even though \s5 fields are not valid USFM
            || ourErrorMessage.startsWith('Expected "f*", "+"') // Might neeed a OHM schema fix?
        )
            adjustedPriority = 294;

        ourErrorObject = {
            priority: adjustedPriority, message: `USFMGrammar: ${ourErrorMessage}`,
            filename,
            characterIndex, excerpt,
            location: givenLocation
        };

        // Save our line number
        if (lineNumberString && lineNumberString.length) {
            // ourErrorObject.lineNumber = Number(lineNumberString);
            //  but we need a temporary fix for the BCS bug which doesn’t include blank lines in the count
            let lineNumber = Number(lineNumberString)
            let notified = false;
            const lines = fileText.split('\n');
            for (let n = 1; n <= lines.length; n++) {
                if (n >= lineNumber) break; // Gone far enough
                if (!lines[n - 1]) {
                    lineNumber += 1; // Increment error line number for each blank line
                    if (!notified) {
                        userLog("Temporarily adjusting BCS grammar error line number to account for blank lines");
                        notified = true;
                    }
                }
            }
            ourErrorObject.lineNumber = lineNumber;
        }
    }

    // debugLog(`  Warnings: ${JSON.stringify(parseWarnings)}`);
    let ourWarnings = [];
    for (const warningString of parseWarnings) {
        // debugLog(`warningString: '${warningString}'`);
        // Clean up their warnings a little: Remove trailing spaces and periods
        let adjustedString = warningString.trim(); // Removes the trailing space
        if (adjustedString.endsWith('.')) adjustedString = adjustedString.slice(0, adjustedString.length - 1);
        ourWarnings.push(adjustedString);
    }

    return { isValidUSFM: !parseError, error: ourErrorObject, warnings: ourWarnings };
}
// end of runBCSGrammarCheck function


export function checkUSFMGrammar(bookID, strictnessString, filename, givenText, givenLocation, checkingOptions) {
    /*
    This function is only used for the demonstration pages -- not for the core!

    bookID is a three-character UPPERCASE USFM book identifier.

    filename parameter can be an empty string if we don’t have one.

     Returns a result object containing a successList and a noticeList
     */
    functionLog(`checkUSFMGrammar(${givenText.length.toLocaleString()} chars, '${givenLocation}')…`);
    //parameterAssert(strictnessString === 'strict' || strictnessString === 'relaxed', `Unexpected strictnessString='${strictnessString}'`);

    let ourLocation = givenLocation;
    if (ourLocation && ourLocation[0] !== ' ') ourLocation = ` ${ourLocation}`;


    const cugResult = { successList: [], noticeList: [] };

    /**
     *
     * @param {string} successString
     */
    function addSuccessMessage(successString) {
        // functionLog(`checkUSFMGrammar success: ${successString}`);
        cugResult.successList.push(successString);
    }
    /**
     *
     * @description - adds a new notice entry from the partial fields given -- adding bookID and filename to the given fields
     * @param {Object} incompleteNoticeObject expected to contain priority, message, characterIndex, exerpt, location
     */
    function addNoticePartial(incompleteNoticeObject) {
        // functionLog(`checkUSFMGrammar notice: (priority=${priority}) ${message}${characterIndex > 0 ? ` (at character ${characterIndex})` : ""}${excerpt ? ` ${excerpt}` : ""}${location}`);
        //parameterAssert(incompleteNoticeObject.priority !== undefined, "cUSFMgr addNoticePartial: 'priority' parameter should be defined");
        //parameterAssert(typeof incompleteNoticeObject.priority === 'number', `cUSFMgr addNoticePartial: 'priority' parameter should be a number not a '${typeof incompleteNoticeObject.priority}': ${incompleteNoticeObject.priority}`);
        //parameterAssert(incompleteNoticeObject.message !== undefined, "cUSFMgr addNoticePartial: 'message' parameter should be defined");
        //parameterAssert(typeof incompleteNoticeObject.message === 'string', `cUSFMgr addNoticePartial: 'message' parameter should be a string not a '${typeof incompleteNoticeObject.message}': ${incompleteNoticeObject.message}`);
        // parameterAssert(characterIndex !== undefined, "cUSFMgr addNoticePartial: 'characterIndex' parameter should be defined");
        if (incompleteNoticeObject.characterIndex) { parameterAssert(typeof incompleteNoticeObject.characterIndex === 'number', `cUSFMgr addNoticePartial: 'characterIndex' parameter should be a number not a '${typeof incompleteNoticeObject.characterIndex}': ${incompleteNoticeObject.characterIndex}`);
        }
        // parameterAssert(excerpt !== undefined, "cUSFMgr addNoticePartial: 'excerpt' parameter should be defined");
        if (incompleteNoticeObject.excerpt) { parameterAssert(typeof incompleteNoticeObject.excerpt === 'string', `cUSFMgr addNoticePartial: 'excerpt' parameter should be a string not a '${typeof excerpt}': ${incompleteNoticeObject.excerpt}`);
        }
        //parameterAssert(incompleteNoticeObject.location !== undefined, "cUSFMgr addNoticePartial: 'location' parameter should be defined");
        //parameterAssert(typeof incompleteNoticeObject.location === 'string', `cUSFMgr addNoticePartial: 'location' parameter should be a string not a '${typeof incompleteNoticeObject.location}': ${incompleteNoticeObject.location}`);
        cugResult.noticeList.push({ ...incompleteNoticeObject, bookID, filename });
    }


    // Main code for checkUSFMGrammar function
    if (books.isExtraBookID(bookID)) // doesn’t work for these
        return cugResult;

    const grammarCheckResult = runBCSGrammarCheck(strictnessString, bookID, givenText, filename, ourLocation, checkingOptions);
    // debugLog(`grammarCheckResult=${JSON.stringify(grammarCheckResult)}`);

    if (!grammarCheckResult.isValidUSFM)
        addNoticePartial({ priority: 944, message: `USFM3 Grammar Check (${strictnessString} mode) doesn’t pass`, filename, location: ourLocation });

    // We only get one error if it fails
    if (grammarCheckResult.error && grammarCheckResult.priority)
        addNoticePartial(grammarCheckResult.error);

    // Display these warnings but with a lowish priority
    for (const warningString of grammarCheckResult.warnings)
        addNoticePartial({ priority: 101, message: `USFMGrammar: ${warningString}`, filename, location: ourLocation });

    addSuccessMessage(`Checked USFM Grammar (${strictnessString} mode) ${grammarCheckResult.isValidUSFM ? "without errors" : " (but the USFM DIDN’T validate)"}`);
    // debugLog(`  checkUSFMGrammar returning with ${result.successList.length.toLocaleString()} success(es) and ${result.noticeList.length.toLocaleString()} notice(s).`);
    // functionLog(`checkUSFMGrammar result is ${JSON.stringify(result)}`);
    return cugResult;
}
// end of checkUSFMGrammar function
