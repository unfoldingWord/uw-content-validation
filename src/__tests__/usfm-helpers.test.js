/* eslint-env jest */

import { extractTextFromComplexUSFM } from '../core/usfm-helpers';

// These are just very short tests to make sure that the USFM helper functions run


describe('check_USFM_Helpers() - ', () => {
  it('Should pass', async () => { // Don't know how to change from async
    const usfmSnippet = `\\ts\\*
\\v 25 \\zaln-s |x-strong="G20680" x-lemma="ἐσθίω" x-morph="Gr,V,MPA2,,P," x-occurrence="1" x-occurrences="1" x-content="ἐσθίετε"\\*\\w Eat|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G39560" x-lemma="πᾶς" x-morph="Gr,RI,,,,ANS," x-occurrence="1" x-occurrences="1" x-content="πᾶν"\\*\\w everything|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G35880" x-lemma="ὁ" x-morph="Gr,RD,,,,ANS," x-occurrence="1" x-occurrences="1" x-content="τὸ"\\*\\w that|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G44530" x-lemma="πωλέω" x-morph="Gr,V,PPP,ANS," x-occurrence="1" x-occurrences="1" x-content="πωλούμενον"\\*\\w is|x-occurrence="1" x-occurrences="1"\\w*
\\w sold|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G17220" x-lemma="ἐν" x-morph="Gr,P,,,,,D,,," x-occurrence="1" x-occurrences="1" x-content="ἐν"\\*\\w in|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G31110" x-lemma="μάκελλον" x-morph="Gr,N,,,,,DNS," x-occurrence="1" x-occurrences="1" x-content="μακέλλῳ"\\*\\w the|x-occurrence="1" x-occurrences="3"\\w*
\\w market|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G33670" x-lemma="μηδείς" x-morph="Gr,RI,,,,ANS," x-occurrence="1" x-occurrences="1" x-content="μηδὲν"\\*\\w without|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G03500" x-lemma="ἀνακρίνω" x-morph="Gr,V,PPA,NMP," x-occurrence="1" x-occurrences="1" x-content="ἀνακρίνοντες"\\*\\w asking|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G12230" x-lemma="διά" x-morph="Gr,P,,,,,A,,," x-occurrence="1" x-occurrences="1" x-content="διὰ"\\*\\w for|x-occurrence="1" x-occurrences="1"\\w*
\\w the|x-occurrence="2" x-occurrences="3"\\w*
\\w sake|x-occurrence="1" x-occurrences="1"\\w*
\\w of|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G35880" x-lemma="ὁ" x-morph="Gr,EA,,,,AFS," x-occurrence="1" x-occurrences="1" x-content="τὴν"\\*\\w the|x-occurrence="3" x-occurrences="3"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G48930" x-lemma="συνείδησις" x-morph="Gr,N,,,,,AFS," x-occurrence="1" x-occurrences="1" x-content="συνείδησιν"\\*\\w conscience|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*.
\\v 26 \\zaln-s |x-strong="G10630" x-lemma="γάρ" x-morph="Gr,CC,,,,,,,," x-occurrence="1" x-occurrences="1" x-content="γὰρ"\\*\\w For|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\* “\\zaln-s |x-strong="G35880" x-lemma="ὁ" x-morph="Gr,EA,,,,NFS," x-occurrence="1" x-occurrences="1" x-content="ἡ"\\*\\w the|x-occurrence="1" x-occurrences="3"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G10930" x-lemma="γῆ" x-morph="Gr,N,,,,,NFS," x-occurrence="1" x-occurrences="1" x-content="γῆ"\\*\\w earth|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\* {\\w is|x-occurrence="1" x-occurrences="1"\\w*}
\\zaln-s |x-strong="G35880" x-lemma="ὁ" x-morph="Gr,EA,,,,GMS," x-occurrence="1" x-occurrences="1" x-content="τοῦ"\\*\\w the|x-occurrence="2" x-occurrences="3"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G29620" x-lemma="κύριος" x-morph="Gr,N,,,,,GMS," x-occurrence="1" x-occurrences="1" x-content="Κυρίου"\\*\\w Lord|x-occurrence="1" x-occurrences="1"\\w*’\\w s|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*,
\\zaln-s |x-strong="G25320" x-lemma="καί" x-morph="Gr,CC,,,,,,,," x-occurrence="1" x-occurrences="1" x-content="καὶ"\\*\\w and|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G35880" x-lemma="ὁ" x-morph="Gr,EA,,,,NNS," x-occurrence="1" x-occurrences="1" x-content="τὸ"\\*\\w the|x-occurrence="3" x-occurrences="3"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G41380" x-lemma="πλήρωμα" x-morph="Gr,N,,,,,NNS," x-occurrence="1" x-occurrences="1" x-content="πλήρωμα"\\*\\w fullness|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*
\\zaln-s |x-strong="G08460" x-lemma="αὐτός" x-morph="Gr,RP,,,3GFS," x-occurrence="1" x-occurrences="1" x-content="αὐτῆς"\\*\\w of|x-occurrence="1" x-occurrences="1"\\w*
\\w it|x-occurrence="1" x-occurrences="1"\\w*\\zaln-e\\*.”
`;
    const rawReextractedUSFM = await extractTextFromComplexUSFM(usfmSnippet);
    // console.log(`Got ${rawReextractedUSFM}`);
    expect(rawReextractedUSFM.length).toEqual(150);
    expect(rawReextractedUSFM).toMatchSnapshot();
  }, 100); // Allow 0.1 seconds

})
