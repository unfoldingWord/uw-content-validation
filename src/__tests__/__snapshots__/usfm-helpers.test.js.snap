// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check_USFM_Helpers() -  Should pass 1`] = `
"‹v25› Eat everything that is sold in the market without asking for the sake of the conscience. For “the earth is the Lord’s, and the fullness of it.”
"
`;
